File: CPP/Algorithm/Comparator.cpp
// Returns: true / false
// Changes the behaviour of the built-in sort() function
// Calling: sort(v.begin(), v.end(), cmp);

// Sorts a vector of pair in ASCENDING order according to the second element:
bool cmp(const pair<int64_t, int64_t> &a, const pair<int64_t, int64_t> &b) {
    return (a.second < b.second);
}

// Sorts a set / multiset of pair in ASCENDING order according to the second element:
struct cmp {
    bool operator() (const pair<int64_t, int64_t> &a, const pair<int64_t, int64_t> &b) const {
        return a.second < b.second;
    }
};

// Sorts a vector of pair in DESCENDING order according to the second element:
bool cmp(const pair<int64_t, int64_t> &a, const pair<int64_t, int64_t> &b) {
    return (a.second > b.second);
}

// Sorts a vector of string in ASCENDING order according to the size():
bool cmp(const string &a, const string &b) {
    return (a.size() < b.size());
}

// Sorts a vector of string in DESCENDING order according to the size():
bool cmp(const string &a, const string &b) {
    return (a.size() > b.size());
}

// Sorts a vector of pair in ASCENDING order according to the sum of the pair:
bool cmp(const pair<int64_t, int64_t> &a, const pair<int64_t, int64_t> &b) {
    return (a.first + a.second) < (b.first + b.second);
}

// Sorts Ascending pair of first elements & Descending pair of second elements
bool cmp(const pair<int64_t, int64_t> &a, const pair<int64_t, int64_t> &b) {
    if(a.first != b.first) return a.first < b.first;
    return a.second > b.second;
}

File: Python/Algorithm/Comparator.py
# Returns: true/false
# Work: Changes the behaviour of the built-in sort() function

def cmp(item):
    return (item[0], -item[1])
pairs = [(1, 2), (1, 3), (2, 1)]

# Work: Prints Ascending pair of first elements & Descending pair of second elements

sorted_pairs = sorted(pairs, key = cmp)

# Work: Sorts a vector of pair in ASCENDING order according to the second element:

sorted_pairs = sorted(pairs, key=lambda x: x[1])

# Work: Sorts a vector of pair in DESCENDING order according to the second element:

sorted_pairs = sorted(pairs, key=lambda x: x[1], reverse=True)

# Work: Sorts a vector of string in ASCENDING order according to the size():

strings = ["abc", "a", "abcd"]
sorted_strings = sorted(strings, key=lambda x: len(x))

# Work: Sorts a vector of string in DESCENDING order according to the size():

sorted_strings = sorted(strings, key=lambda x: len(x), reverse=True)